rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+index: create the index, copy the data into the index, then issue the command to
+point the alias to the new index. This process is encapsulated in two actions:
+the `create_and_populate_filtered_index` action described above, and the
+existing `point_alias` action.
+
+> **Note**
+>
+> While both processes have a "copy" step, they are notably different. The
+> origin index receives its documents from Postgres via ingestion workers, which
+> push the documents into Elasticsearch. The filtered index receives its
+> documents via the reindex API and does not require the ingestion workers.
+
+### Airflow data refresh DAG factory (overview step 2)
+
+Adding these new steps to the Airflow data refresh DAG is simple and only
+requires adding the two additional actions to the
+[existing `action_data_map`](https://github.com/WordPress/openverse-catalog/blob/4b5811aa240ba3ee2d92adea3a8abf88eb1e8aa4/openverse_catalog/dags/data_refresh/data_refresh_task_factory.py#L182)
+with the target alias configured on the `point_alias` action, just as it is with
+the regular `promote` action.
+
+```py
+action_data_map: dict[str, dict] = {
+    "ingest_upstream": {},
+    "promote": {"alias": target_alias},
+    "create_and_populate_filtered_index": {},
+    "point_alias": {"alias": f"{target_alias}-filtered"}
+}
+```
+
+### Query the filtered index (overview step 3)
+
+We must add a new boolean query parameter to the search query serialiser,
+`unstable__include_sensitive_results`. This parameter will default to `False`.
+We will remove the `unstable` designation during the wrap-up of this
+implementation plan. This parameter should also reflect the state of the
+`mature` parameter: when `mature` is `True`, the new parameter should also be
+`True`. This prepares us to deprecate the `mature` parameter when we remove the
+`unstable` designation from the new parameter.
+
+When `unstable__include_sensitive_results` is `False` (default behaviour), query
+the filtered index by appending the `-filtered` suffix to the index name. Update
+the existing `mature` parameter handling to check
+`unstable__include_sensitive_results`. When the new parameter is `False`,
+results with the `mature` set to `true` should also be filtered out.
+
+When `unstable__include_sensitive_results` is `True`, query the origin index
+(existing behaviour).
+
+### Derive the list of sensitive document IDs based on the filtered index (overview step 4)
+
+When `unstable__include_sensitive_results` is `True`, we need to derive which
+documents in the results have sensitive textual content. To do this, we will
+rely on the fact that those documents are _not_ in the filtered index.
+Therefore, we can pull the list of `_id`s for the documents retrieved from the
+origin index and query for their presence in the filtered index. If they are in
+the filtered index, then we know they do not have sensitive textual content.
+Something along the following lines of pseudocode:
```

Comment:
This is very much a nit, but there's some oscillation between `True`/`true` and `False`/`false` here - is there any importance behind that difference? If not, would it make sense to make them consistent across the document?

Reply:
No reason other than missing the shift bar a couple times :slightly_smiling_face: I'll consolidate to using the Python title cased version.

Reply:
Ah, actually, the difference between `true` and `True` here _was_ intentional! In Elasticsearch the "mature" property is a JSON field, so the value is indeed `true` with a lowercase T. I'm still changing the language to not reference the specific values, but I just wanted to note that there is indeed a small difference here due to the data representation model that Elasticsearch is actually using.